src/webgpu/api/operation/vertex_state.spec.ts

```diff
@@ -0,0 +1,348 @@
+// than the size required by triangle list, otherwise it also render nothing.
+g.test('index_format_uint16')
+  .params(params().combine(poptions('indexOffset', [0, 6, 18])))
+  .fn(t => {
+    const { indexOffset } = t.params;
+
+    const indexArray = new Uint16Array([1, 2, 0, 0, 0, 0, 0, 1, 3, 0]);
+    const indexBuffer = t.CreateIndexBuffer(indexArray, 'uint16');
+    const indexCount = indexArray.length;
+    t.run(indexBuffer, indexCount, 'uint16', indexOffset);
+
+    let expectedShape: RenderShape;
+    if (indexOffset === 0) {
+      expectedShape = RenderShape.Square;
+    } else if (indexOffset === 6) {
+      expectedShape = RenderShape.BottomLeftTriangle;
+    } else {
+      expectedShape = RenderShape.Nothing;
+    }
+
+    const expectedTextureValues = t.CreateExpectedUint8Array(expectedShape);
+    t.expectContents(t.result, expectedTextureValues);
+  });
+
+// Test indexing draw with index format of uint32. If this is interpreted as uint16, then it would
+// be 0, 1, 0, ... and would draw nothing. And the index buffer size - offset must be not less than
+// the size required by triangle list, otherwise it also render nothing.
+g.test('index_format_uint32')
+  .params(params().combine(poptions('indexOffset', [0, 12, 36])))
+  .fn(t => {
+    const { indexOffset } = t.params;
+
+    const indexArray = new Uint32Array([1, 2, 0, 0, 0, 0, 0, 1, 3, 0]);
+    const indexBuffer = t.CreateIndexBuffer(indexArray, 'uint32');
+    const indexCount = indexArray.length;
+    t.run(indexBuffer, indexCount, 'uint32', indexOffset);
+
+    let expectedShape: RenderShape;
+    if (indexOffset === 0) {
+      expectedShape = RenderShape.Square;
+    } else if (indexOffset === 12) {
+      expectedShape = RenderShape.BottomLeftTriangle;
+    } else {
+      expectedShape = RenderShape.Nothing;
+    }
+
+    const expectedTextureValues = t.CreateExpectedUint8Array(expectedShape);
+    t.expectContents(t.result, expectedTextureValues);
+  });
+
+// Test primitive restart with each primitive topology. The primitive restart value can used with
+// strip primitive topologies ('line-strip' or 'triangle-strip').
+// For line-strip, [0, 1, PRIM_RESTART, 2, 3] only render a X shape.
+// For triangle-strip, [0, 1, 3, PRIM_RESTART, 2] only the first triangle.
+// Others are expected to be renderred the same as without primitive restart.
+g.test('primitive_restart')
+  .params(
+    params()
+      .combine(poptions('indexFormat', ['uint16', 'uint32']))
+      .combine(
+        poptions('primitiveTopology', [
+          'point-list',
+          'line-list',
+          'line-strip',
+          'triangle-list',
+          'triangle-strip',
+        ])
+      )
+  )
+  .fn(t => {
+    const { indexFormat, primitiveTopology } = t.params;
+
+    let indexArray: Uint16Array | Uint32Array;
+    if (indexFormat === 'uint16') {
+      if (primitiveTopology === 'triangle-list') {
+        indexArray = new Uint16Array([0, 1, 3, 0xffff, 2, 1, 0, 0]);
```

Comment:
If primitive restart can indeed be disabled for lists, 0xffff will go out of bounds of the vertex buffer. Maybe we should make the vertex buffer large enough that they wouldn't. Do we have a strict guarantee on what happens when you go out of bounds of the vertex buffer? I think we allow a range of behaviors, which we shouldn't test here. Of course we can't do this for uint32 but at least we can do it for uint16.

Reply:
I don't clear that what the 0xffff will be used for in non-strip topologies when primitive restart can indeed be disabled for lists. If it's used as a big number in index buffer, I think we need the tests, currently it draws nothing with an index out of bounds of the vertex buffer, no error is thrown.  But if it's checked as an invalid index, and skipped, our tests will cover it. I'm not sure what kind of behavior it will be.

Reply:
> Also note that if ANY index goes out of bounds, I'm pretty sure we allow discarding the entire draw call (not just that primitive).

So for the lists, our expected result should be nothing is renderred, right?

Reply:
I think there are multiple possible expected results: the entire draw could be discarded, just the invalid triangle could be discarded, or the invalid triangle could be rendered by clamping the index (0xFFFF) to the max valid index value (3). This is why it would be a good idea to make the vertex buffer large enough that 0xFFFF does not go out of bounds. We can test out of bounds behavior, including for 0xFFFF, in another test.

Reply:
Another (smaller) reason to make sure 0xFFFF is a valid index in the vertex buffer is that, if an implementation did triangle-strip with primitive restart disabled, they would still pass if just invalid triangles were discarded.